File: quick_sort_step.java
import java.io.*;
import java.util.*;

public class quick_sort_step {

    public static void main(String[] args) throws Exception {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter CSV filename: ");
        String filename = scanner.nextLine();

        System.out.print("Enter start row: ");
        int startRow = scanner.nextInt();

        System.out.print("Enter end row: ");
        int endRow = scanner.nextInt();
        scanner.close();

        List<DataRow> data = readSubset(filename, startRow, endRow);
        List<String> steps = new ArrayList<>();

        steps.add(data.toString());
        quickSort(data, 0, data.size() - 1, steps);

        String outputFile = "quick_sort_step_" + startRow + "_" + endRow + ".txt";
        writeSteps(steps, outputFile);
        System.out.println("Quick sort steps written to " + outputFile);
    }

    public static List<DataRow> readSubset(String filename, int start, int end) throws Exception {
        List<DataRow> list = new ArrayList<>();
        BufferedReader br = new BufferedReader(new FileReader(filename));
        String line;
        int row = 0;
        while ((line = br.readLine()) != null) {
            if (row >= start && row <= end) {
                String[] parts = line.split(",");
                list.add(new DataRow(Long.parseLong(parts[0]), parts[1]));
            }
            row++;
        }
        br.close();
        return list;
    }

    public static void quickSort(List<DataRow> list, int low, int high, List<String> steps) {
        if (low < high) {
            int pi = partition(list, low, high);
            steps.add("pi=" + pi + " " + list.toString());
            quickSort(list, low, pi - 1, steps);
            quickSort(list, pi + 1, high, steps);
        }
    }

    public static int partition(List<DataRow> list, int low, int high) {
        DataRow pivot = list.get(high);
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (list.get(j).compareTo(pivot) <= 0) {
                i++;
                Collections.swap(list, i, j);
            }
        }
        Collections.swap(list, i + 1, high);
        return i + 1;
    }

    public static void writeSteps(List<String> steps, String filename) throws Exception {
        BufferedWriter writer = new BufferedWriter(new FileWriter(filename));
        for (String step : steps) {
            writer.write(step);
            writer.newLine();
        }
        writer.close();
    }
}

class DataRow implements Comparable<DataRow> {
    long number;
    String text;

    public DataRow(long number, String text) {
        this.number = number;
        this.text = text;
    }

    @Override
    public int compareTo(DataRow other) {
        return Long.compare(this.number, other.number);
    }

    @Override
    public String toString() {
        return number + "/" + text;
    }
}

File: quick_sort_step.py
import csv

def read_csv(filename):
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        return [(int(row[0]), row[1]) for row in reader]

def format_data(arr):
    return "[" + ", ".join(f"{num}/{text}" for num, text in arr) + "]"

def quick_sort_step(arr, low, high, steps):
    if low < high:
        pi = partition(arr, low, high, steps)
        quick_sort_step(arr, low, pi - 1, steps)
        quick_sort_step(arr, pi + 1, high, steps)

def partition(arr, low, high, steps):
    pivot = arr[high][0]
    i = low - 1
    for j in range(low, high):
        if arr[j][0] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    
    steps.append(f"pi={i+1} {format_data(arr)}")
    return i + 1

def main():
    input_file = input("Enter dataset filename (e.g., dataset_sample_1000.csv): ").strip()
    data = read_csv(input_file)
    total_rows = len(data)

    print(f"\nDataset loaded. Total rows: {total_rows} (index from 0 to {total_rows - 1})")

    try:
        start_row = int(input(f"Enter start row (0 to {total_rows - 1}): ").strip())
        end_row = int(input(f"Enter end row (0 to {total_rows - 1}): ").strip())
    except ValueError:
        print("Please enter valid integers.")
        return

    if start_row < 0 or end_row >= total_rows or start_row > end_row:
        print(" Invalid row range. Make sure 0 <= start <= end < total rows.")
        return

    selected = data[start_row:end_row + 1]
    print(f"📘 Selected rows: {start_row} to {end_row} ({len(selected)} entries)")

    steps = [format_data(selected)]
    quick_sort_step(selected, 0, len(selected) - 1, steps)

    output_filename = f"quick_sort_step_{start_row}_{end_row}.txt"
    with open(output_filename, 'w') as f:
        for step in steps:
            f.write(step + "\n")

    print(f"Output written to: {output_filename}")

if __name__ == "__main__":
    main()
